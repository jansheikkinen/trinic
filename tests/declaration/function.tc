// Function

function(void) -> void do 1; end
function testA(void) -> void do 2; end
function testB(x: int8) -> void do 3; end
function testC(void) -> int8 do 4; end
function testD(x: int8) -> int8 where x != 69, return == 420 do 5; end
function testE(x: T) -> T for T: trait Add do 6; end
function testF(x: T, y: V) -> T for T + V: trait Add + trait Display do 7; end
function testG(x: T, y: V) -> T
  where x != 1
  for T: trait Add, V: trait Display
do 8; end
function testH(void) -> void do 9; 10; end
