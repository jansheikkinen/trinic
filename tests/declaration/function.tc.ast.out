
### AST GENERATION ###
Declaration FUNCTION 
  Stmt EXPRESSION
    Expr LITERAL
      INT 1
Declaration FUNCTION  testA
  Stmt EXPRESSION
    Expr LITERAL
      INT 2
Declaration FUNCTION  testB
  Arg IDENTIFIER
    Identifier_Arg x
      Type TYPE INT8
  Stmt EXPRESSION
    Expr LITERAL
      INT 3
Declaration FUNCTION  testC
  Type TYPE INT8
  Stmt EXPRESSION
    Expr LITERAL
      INT 4
Declaration FUNCTION  testD
  Arg IDENTIFIER
    Identifier_Arg x
      Type TYPE INT8
  Type TYPE INT8
  Arg EXPR
    Expr BINARY !=
      Expr VARIABLE x
      Expr LITERAL
        INT 69
    Expr BINARY ==
      Expr VARIABLE return
      Expr LITERAL
        INT 420
  Stmt EXPRESSION
    Expr LITERAL
      INT 5
Declaration FUNCTION  testE
  Arg IDENTIFIER
    Identifier_Arg x
      Type IDENTIFIER T
  Type IDENTIFIER T
  Arg GENERIC_DEFS
    Generic_Def
      Arg GENERIC_DEFL
        IDENTIFIER T
      Arg GENERIC_DEFR
        Type TRAIT Add
  Stmt EXPRESSION
    Expr LITERAL
      INT 6
Declaration FUNCTION  testF
  Arg IDENTIFIER
    Identifier_Arg x
      Type IDENTIFIER T
    Identifier_Arg y
      Type IDENTIFIER V
  Type IDENTIFIER T
  Arg GENERIC_DEFS
    Generic_Def
      Arg GENERIC_DEFL
        IDENTIFIER T
        IDENTIFIER V
      Arg GENERIC_DEFR
        Type TRAIT Add
        Type TRAIT Display
  Stmt EXPRESSION
    Expr LITERAL
      INT 7
Declaration FUNCTION  testG
  Arg IDENTIFIER
    Identifier_Arg x
      Type IDENTIFIER T
    Identifier_Arg y
      Type IDENTIFIER V
  Type IDENTIFIER T
  Arg EXPR
    Expr BINARY !=
      Expr VARIABLE x
      Expr LITERAL
        INT 1
  Arg GENERIC_DEFS
    Generic_Def
      Arg GENERIC_DEFL
        IDENTIFIER T
      Arg GENERIC_DEFR
        Type TRAIT Add
    Generic_Def
      Arg GENERIC_DEFL
        IDENTIFIER V
      Arg GENERIC_DEFR
        Type TRAIT Display
  Stmt EXPRESSION
    Expr LITERAL
      INT 8
Declaration FUNCTION  testH
  Stmt EXPRESSION
    Expr LITERAL
      INT 9
  Stmt EXPRESSION
    Expr LITERAL
      INT 10
